File: ios/TyphoonKit/Sources/TyphoonKit/RelaySelector.swift
import Foundation

public struct RelaySelector: Sendable {
    public init() {}

    public func orderedCandidates(from relays: [RelayDescriptor], now: Date = Date()) -> [RelayDescriptor] {
        let usable = relays.filter { $0.isUsable(now: now) }
        return usable.filter { Self.isIPv6Literal($0.publicHost) } +
            usable.filter { !Self.isIPv6Literal($0.publicHost) }
    }

    public func selectFirstUsable(from relays: [RelayDescriptor], now: Date = Date()) -> RelayDescriptor? {
        orderedCandidates(from: relays, now: now).first
    }

    private static func isIPv6Literal(_ host: String) -> Bool {
        let trimmed = host.trimmingCharacters(in: CharacterSet(charactersIn: "[]"))
        return trimmed.contains(":")
    }
}

File: android/app/src/main/java/com/typhoon/client/model/RelaySelector.kt
package com.typhoon.client.model

import java.time.Instant

class RelaySelector {
    fun orderedCandidates(relays: List<RelayDescriptor>, now: Instant): List<RelayDescriptor> {
        val usable = relays.filter { it.isUsable(now) }
        return usable.filter { it.publicHost.isIPv6Literal() } +
            usable.filterNot { it.publicHost.isIPv6Literal() }
    }

    fun selectFirstUsable(relays: List<RelayDescriptor>, now: Instant): RelayDescriptor? =
        orderedCandidates(relays, now).firstOrNull()

    private fun String.isIPv6Literal(): Boolean =
        trim('[', ']').contains(":")
}
